### PR TITLE
feat(events): Update and Add Cognito User Pool Events

### LINKED ISSUE #3987: Feature Request: Update some outdated Cognito user pool event data classes
### Expected Behavior

Accessing code_parameter property of cognito_user_pool_event.CustomMessageTriggerEvent returns the proper encrypted code.

### Current Behaviour

Mapping the CustomMessageTriggerEvent data class of cognito_user_pool_event leads to a KeyError because the received event property names have been changed by AWS.
The new request and documentation can be found at : https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sms-sender.html#custom-sms-sender-parameters 

### Code snippet

```python
from aws_lambda_powertools.utilities import data_classes
from aws_lambda_powertools.utilities.data_classes import cognito_user_pool_event

@data_classes.event_source(data_class=cognito_user_pool_event.CustomMessageTriggerEvent)
def handler(event, context):
    code = event.request.code_parameter
```

### Possible Solution

Update data class source for cognito_user_pool_event to reflect the updated event. Currently it fetches `event['request']['code_parameter']` but it should fetch `event['request']['code']`

If that's OK I could contribute, I'd be glad to go through all the source events for Cognito as I see it's been quite a while since they have been added to the project. I could go through the current AWS docs and test it with existing Cognito Pool event behavior for all the events as we rely heavily on those in our current project. 

For now I've recreated the event class in our local code, deriving `BaseTriggerEvent`.

### Steps to Reproduce

This is the event trigger that is sent by Cognito that can be used as an input to the lambda handler.

```
{
	"version": "1",
	"triggerSource": "CustomSMSSender_ResendCode",
	"region": "<region>",
	"userPoolId": "<pool_id>",
	"userName": "<user_name>",
	"callerContext": {
		"awsSdkVersion": "aws-sdk-unknown-unknown",
		"clientId": "<client_id>"
	},
	"request": {
		"type": "customEmailSenderRequestV1",
		"code": "<encrypted_code>",
		"clientMetadata": {},
		"userAttributes": {
			<usual_attributes>
		}
	}
}
```

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.11

### Packaging format used

PyPi

### Debugging logs

```python
[ERROR] KeyError: 'codeParameter'
Traceback (most recent call last):
  File "/opt/python/aws_lambda_powertools/logging/logger.py", line 447, in decorate
    return lambda_handler(event, context, *args, **kwargs)
  File "/opt/python/aws_lambda_powertools/middleware_factory/factory.py", line 135, in wrapper
    response = middleware()
  File "/opt/python/aws_lambda_powertools/utilities/data_classes/event_source.py", line 39, in event_source
    return handler(data_class(event), context)
  File "/var/task/custom_sms_trigger/index.py", line 39, in handler
    return handle_sms_event(event=event)
  File "/var/task/custom_sms_trigger/index.py", line 25, in handle_sms_event
    if not event.request.code_parameter:
  File "/opt/python/aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py", line 271, in code_parameter
    return self["request"]["codeParameter"]
```

### PR DESCRIPTION
**Pull request by knightmre**
**Issue number:** [#3987](https://github.com/aws-powertools/powertools-lambda-python/issues/3987)

### Changes

> Please provide a summary of what's being changed

This change updates the UserMigrationTriggerEvent data class, adds the V2 of [Pre Token Generation Lambda Triggers](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html) and adds two new data classes for [Custom Sender Lambda triggers](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sender-triggers.html)

### User experience

> Please share what the user experience looks like before and after this change

Events did not exist before. A workaround was to self-implement the data classes by inheriting base classes like so:

**Before**

```python
from typing import Optional

from aws_lambda_powertools.utilities.data_classes import cognito_user_pool_event
from aws_lambda_powertools.utilities.data_classes import common


class CustomSMSSenderTriggerEventRequest(common.DictWrapper):
    @property
    def code(self) -> str:
        """A string for you to use as the placeholder for the verification code in the custom message."""
        return self["request"]["code"]

    @property
    def user_attributes(self) -> dict[str, str]:
        """One or more name-value pairs representing user attributes. The attribute names are the keys."""
        return self["request"]["userAttributes"]

    @property
    def client_metadata(self) -> Optional[dict[str, str]]:
        """One or more key-value pairs that you can provide as custom input to the Lambda function
        that you specify for the pre sign-up trigger."""
        return self["request"].get("clientMetadata")


class CustomSMSSenderEvent(cognito_user_pool_event.BaseTriggerEvent):

    @property
    def request(self) -> CustomSMSSenderTriggerEventRequest:
        return CustomSMSSenderTriggerEventRequest(self._data)

```

```python

from . import CustomSMSSenderEvent

@data_classes.event_source(data_class=CustomSMSSenderEvent)
def handler(event: CustomSMSSenderEvent, context):
    return handle_event(event=event)
```

**After**

```python

from aws_lambda_powertools.utilities.data_classes.cognito_user_pool_event import CustomSMSSenderTriggerEvent

@data_classes.event_source(data_class=CustomSMSSenderTriggerEvent)
def handler(event: CustomSMSSenderTriggerEvent, context):
    return handle_event(event=event)
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [ ] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Hi @knightmre this is a super detailed PR with an impressive level of awesome docs/documentation. I'm working on other PRs and will schedule the first review of this PR for early next week.

Thanks

**Comment by knightmre** [comment]
> Hi @knightmre this is a super detailed PR with an impressive level of awesome docs/documentation. I'm working on other PRs and will schedule the first review of this PR for early next week.
> 
> Thanks

Hi, thanks for your time, I just found out I missed out on some tests so I'll be reworking this a little before your review.

**Comment by leandrodamascena** [comment]
Reviewing this now.

File: aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py [hunk 1]
```diff
@@ -520,6 +547,123 @@ def set_group_configuration_preferred_role(self, value: str):
         self["groupOverrideDetails"]["preferredRole"] = value
 
 
+class TokenClaimsAndScopeOverrideDetails(DictWrapper):
+    @property
+    def claims_to_add_or_override(self) -> Optional[Dict[str, str]]:
+        return self.get("claimsToAddOrOverride")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py]:
```suggestion
    def claims_to_add_or_override(self) -> Dict[str, str]:
        return self.get("claimsToAddOrOverride") or {}
```

File: aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py [hunk 2]
```diff
@@ -520,6 +547,123 @@ def set_group_configuration_preferred_role(self, value: str):
...
+
+    @claims_to_add_or_override.setter
+    def claims_to_add_or_override(self, value: Dict[str, str]):
+        """A map of one or more key-value pairs of claims to add or override.
+        For group related claims, use groupOverrideDetails instead."""
+        self._data["claimsToAddOrOverride"] = value
+
+    @property
+    def claims_to_suppress(self) -> Optional[List[str]]:
+        return self.get("claimsToSuppress")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py]:
```suggestion
    def claims_to_suppress(self) -> List[str]:
        return self.get("claimsToSuppress") or []
```

File: aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py [hunk 3]
```diff
@@ -520,6 +547,123 @@ def set_group_configuration_preferred_role(self, value: str):
...
+
+    @claims_to_suppress.setter
+    def claims_to_suppress(self, value: List[str]):
+        """A list that contains claims to be suppressed from the identity token."""
+        self._data["claimsToSuppress"] = value
+
+    @property
+    def scopes_to_add(self) -> Optional[List[str]]:
+        return self.get("scopesToAdd")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py]:
```suggestion
    def scopes_to_add(self) -> List[str]:
        return self.get("scopesToAdd") or []
```

File: aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py [hunk 4]
```diff
@@ -520,6 +547,123 @@ def set_group_configuration_preferred_role(self, value: str):
...
+
+    @scopes_to_add.setter
+    def scopes_to_add(self, value: List[str]):
+        self._data["scopesToAdd"] = value
+
+    @property
+    def scopes_to_suppress(self) -> Optional[List[str]]:
+        return self.get("scopesToSuppress")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py]:
```suggestion
    def scopes_to_suppress(self) -> List[str]:
        return self.get("scopesToSuppress") or []
```

File: aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py [hunk 5]
```diff
@@ -822,3 +1007,77 @@ def request(self) -> VerifyAuthChallengeResponseTriggerEventRequest:
...
+    @property
+    def client_metadata(self) -> Optional[Dict[str, str]]:
+        """One or more key-value pairs that you can provide as custom input to the
+        custom email sender Lambda function trigger. To pass this data to your Lambda function,
+        you can use the ClientMetadata parameter in the AdminRespondToAuthChallenge and
+        RespondToAuthChallenge API actions. Amazon Cognito doesn't include data from the
+        ClientMetadata parameter in AdminInitiateAuth and InitiateAuth API operations
+        in the request that it passes to the post authentication function.
+        """
+        return self["request"].get("clientMetadata")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py]:
```suggestion
    def client_metadata(self) -> Dict[str, str]:
        """One or more key-value pairs that you can provide as custom input to the
        custom email sender Lambda function trigger. To pass this data to your Lambda function,
        you can use the ClientMetadata parameter in the AdminRespondToAuthChallenge and
        RespondToAuthChallenge API actions. Amazon Cognito doesn't include data from the
        ClientMetadata parameter in AdminInitiateAuth and InitiateAuth API operations
        in the request that it passes to the post authentication function.
        """
        return self["request"].get("clientMetadata") or {}
```

File: aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py [hunk 6]
```diff
@@ -822,3 +1007,77 @@ def request(self) -> VerifyAuthChallengeResponseTriggerEventRequest:
...
+    @property
+    def client_metadata(self) -> Optional[Dict[str, str]]:
+        """One or more key-value pairs that you can provide as custom input to the
+        custom SMS sender Lambda function trigger. To pass this data to your Lambda function,
+        you can use the ClientMetadata parameter in the AdminRespondToAuthChallenge and
+        RespondToAuthChallenge API actions. Amazon Cognito doesn't include data from the
+        ClientMetadata parameter in AdminInitiateAuth and InitiateAuth API operations
+        in the request that it passes to the post authentication function.
+        """
+        return self["request"].get("clientMetadata")
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cognito_user_pool_event.py]:
```suggestion
    def client_metadata(self) -> Dict[str, str]:
        """One or more key-value pairs that you can provide as custom input to the
        custom SMS sender Lambda function trigger. To pass this data to your Lambda function,
        you can use the ClientMetadata parameter in the AdminRespondToAuthChallenge and
        RespondToAuthChallenge API actions. Amazon Cognito doesn't include data from the
        ClientMetadata parameter in AdminInitiateAuth and InitiateAuth API operations
        in the request that it passes to the post authentication function.
        """
        return self["request"].get("clientMetadata") or {}
```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @knightmre! Well done! This was one of the most detailed PRs I've looked at. Creating all those events in Cognito must have taken a lot of hard work and I'm sure you spent some days. I tried making some myself, and it took me over a day. 

I just made a few small comments to your great work. 🚀

**Comment by knightmre** [comment]
> Hey @knightmre! Well done! This was one of the most detailed PRs I've looked at. Creating all those events in Cognito must have taken a lot of hard work and I'm sure you spent some days. I tried making some myself, and it took me over a day.
> 
> I just made a few small comments to your great work. 🚀

Hi @leandrodamascena Thanks for the review. I've included your comments and updated the PR.

**Review by leandrodamascena — APPROVED** [review]
I'm going to say this again: amazing work @knightmre!

APPROVED!

**Review by sthulb — APPROVED** [review]

**Pull request merged**